### DISCUSSION TITLE
running odin commands via docker

### DISCUSSION THREAD

**Discussion by kpengwin** [mbround18/valheim-docker#99] [Q&A] [answered]
This is likely me being unfamiliar with docker, but I'm having some trouble running odin commands via docker exec.

Having stopped the server with `kill -2 $pid` I would like to restart it via a script, so I've tried running something like:
`docker exec --user steam -it $VALHEIM_CONTAINER /path/to/odin start`
but this gives output like the following:
```[ODIN][INFO]  - Setting up start scripts...
[ODIN][INFO]  - Looking for burial mounds...
[ODIN][INFO]  - Successfully written /home/steam/steamcmd/start_server_rusty.sh
[ODIN][INFO]  - Created the /home/steam/steamcmd/start_server_rusty.sh script successfully!
[ODIN][INFO]  - Successfully set /home/steam/steamcmd/start_server_rusty.sh to executable
[ODIN][INFO]  - Successfully set /home/steam/steamcmd/start_server_rusty.sh to executable!
[ODIN][ERROR] - Could not find server executable! Please install the server!
```
however if i run:
`docker exec --user steam -it $VALHEIM_CONTAINER bash`
followed by
`odin start`
it works:
```steam@00210c0e6b61:~/valheim$ odin start
[ODIN][INFO]  - Setting up start scripts...
[ODIN][INFO]  - Looking for burial mounds...
[ODIN][INFO]  - Successfully written /home/steam/valheim/start_server_rusty.sh
[ODIN][INFO]  - Created the /home/steam/valheim/start_server_rusty.sh script successfully!
[ODIN][INFO]  - Successfully set /home/steam/valheim/start_server_rusty.sh to executable
[ODIN][INFO]  - Successfully set /home/steam/valheim/start_server_rusty.sh to executable!
[ODIN][INFO]  - Exit with code exit code: 0
[ODIN][INFO]  - Server has started...
[ODIN][INFO]  - Check out ./output.log for the logs.
[ODIN][INFO]  - Keep an eye out for "Game server connected" and you server should be live!
```
This seems like a working directory problem, since the server is in fact located in /home/steam/valheim/ not /home/steam/steamcmd/, but I'm not sure what to pass in to fix it. Specifically I notice that there's a version of the start_server_rusty.sh script in both folders, and both include an absolute path to the server (correct for the one in /home/steam/valheim, incorrect for the other) so that's the source of the error, but I'm not sure why it is defaulting to the wrong one.

I intend to keep poking around until i find a way to make this work (probably whatever you have to do to pass in a working directory to your command), but I'd appreciate any tips here. (And, it seems like it would be nice if it defaulted to working correctly.)

**Comment by CosmicHorrorDev** [comment, marked as answer]
It looks like this is a potentially old docker image. I'd wait for @mbround18 to weigh in, but `rusty_startup_script.sh` is no longer used in newer versions of odin.

**Reply by kpengwin** [reply]
TY, good to know.

**Comment by mbround18** [comment]
Because this looks like an older version of the image, I would recommend running odin commands while inside the container as the user due to some jenky environment variables. 



These have been resolved on the `:latest` and `:1.2.0` versions of the docker image. I recommend you pull a newer version of the image as it provides a much cleaner interface especially with odin now in the systems path of `/usr/local/bin/odin`

**Reply by kpengwin** [reply]
I love how "older" means like 3 days ago haha. Thanks, I'll pull a newer version.

**Reply by mbround18** [reply]
> I love how "older" means like 3 days ago haha. Thanks, I'll pull a newer version.

lol well in full disclosure this is my first time coding with Rust and there was a bit of a learning curve :P I know rust can go low level but the tools provided in base rust were not letting me. However, I found a fantastic plugin called Daemonize which allowed me to go much lower level and run the process in a detached manor. <3 thats where that new methodology of launching the server came from :3  since 1.2.+ it no longer drops down to shell to launch the server and is all handled internally.

**Reply by kpengwin** [reply]
Nice - It seems like a good learning project! Rust is probably something I'll pick up at some point, it's an interesting language. Really appreciate the work you've put in here, hosting the server with your container has worked extremely well for me.
